### ISSUE TITLE
EXC_BAD_ACCESS error in xcode 8 when trying and purchase a product

### ISSUE THREAD

**Issue by bboldi** [j3k0/cordova-plugin-purchase#537]
When I'm trying to use the plugin in my ionic 1 application xcode stops and gives me the following error:

`Thread 1: EXC_BAD_ACCESS (code=1, address=0x79869170)`

the brakepoing is in the file `inAppPurchase.m` line 364

> `[[SKPaymentQueue defaultQueue] addPayment:payment];`

here is the whole function:

```
- (void) purchase: (CDVInvokedUrlCommand*)command
{
    DLog(@"About to do IAP");
    id identifier = [command.arguments objectAtIndex:0];
    id quantity =   [command.arguments objectAtIndex:1];

    SKMutablePayment *payment = [SKMutablePayment paymentWithProduct:[self.list objectForKey:identifier]];
    if ([quantity respondsToSelector:@selector(integerValue)]) {
        payment.quantity = [quantity integerValue];
    }
    [[SKPaymentQueue defaultQueue] addPayment:payment];
}
```

this is what i have in xcode console window:

```
2016-12-13 17:48:28.526112 app[415:34889] THREAD WARNING: ['InAppPurchase'] took '13.864990' ms. Plugin should use a background thread.

2016-12-13 17:48:28.586510 app[415:34889] InAppPurchase[objc]: Getting products data

2016-12-13 17:48:28.586721 app[415:34889] InAppPurchase[objc]: Set has 1 elements

2016-12-13 17:48:28.586779 app[415:34889] InAppPurchase[objc]:  - xxxx

2016-12-13 17:48:28.586941 app[415:34889] InAppPurchase[objc]: Starting product request...

2016-12-13 17:48:28.587193 app[415:34889] InAppPurchase[objc]: Product request started

2016-12-13 17:48:29.495091 app[415:34889] InAppPurchase[objc]: productsRequest: didReceiveResponse:

2016-12-13 17:48:29.495206 app[415:34889] InAppPurchase[objc]: Has 1 validProducts

2016-12-13 17:48:29.497921 app[415:34889] InAppPurchase[objc]:  - xxxx: This is a test subscription

2016-12-13 17:48:29.501294 app[415:34889] InAppPurchase[objc]: productsRequest: didReceiveResponse: sendPluginResult: (

        (

                {

            currency = USD;

            description = "Just a test subscription description for sandbox";

            id = "xxxx";

            price = "$4.99";

            title = "This is a test subscription";

        }

    ),

        (

    )

)

2016-12-13 17:48:30.965666 app[415:34889] InAppPurchase[objc]: About to do IAP
```

I'm using `ios 4.3.1` platform ... with xcode 8.2 (it's the same with any xcode 8 version)

**Comment by bboldi** [comment]
I have been trying to solve this with no success so far. I tried to use different cordova versions, different cordova-ios versions and different versions of the plugin, the result is the same - the app crashes. The most frustrating thing is, sometimes it works, sometimes (most of the times) it does not ... and there's a pattern (sort of) when I run the app in XCode it usually works the first time, then if i kill the app, and start it from the phone it dies, then if I go to multitasking and click on the app again then it works ... the app dies right before (or maybe during) the purchase dialog appears. I have tried to create a totally new application, and reproduce the problem, could not do it - it showed the purchase dialog every time. Here is the plugin list, 

```
cc.fovea.cordova.purchase 6.0.0 "Purchase"
com.verso.cordova.clipboard 0.1.0 "Clipboard"
cordova-custom-config 3.1.2 "cordova-custom-config"
cordova-plugin-camera 2.3.1 "Camera"
cordova-plugin-compat 1.1.0 "Compat"
cordova-plugin-device 1.1.4 "Device"
cordova-plugin-email 1.2.2 "EmailComposer"
cordova-plugin-facebook4 1.7.4 "Facebook Connect"
cordova-plugin-file 4.3.1 "File"
cordova-plugin-file-transfer 1.6.1 "File Transfer"
cordova-plugin-geolocation 2.4.1 "Geolocation"
cordova-plugin-inappbrowser 1.5.0 "InAppBrowser"
cordova-plugin-nativeaudio 3.0.7 "Cordova Native Audio"
cordova-plugin-splashscreen 4.0.1 "Splashscreen"
cordova-plugin-sqlite-2 1.0.4 "SQLitePlugin"
cordova-plugin-whitelist 1.3.1 "Whitelist"
cordova.plugins.diagnostic 3.3.2 "Diagnostic"
cordova_app_version_plugin 0.2.6 "getAppVersion"
ionic-plugin-deploy 0.6.5 "IonicDeploy"
ionic-plugin-keyboard 2.2.1 "Keyboard"
onesignal-cordova-plugin 2.0.10 "OneSignal Push Notifications"
```

maybe it has something to do with that ... the more I'm trying to debug this, the more question I have. Any tip, idea help with this issue would be much appreciated.

**Comment by pedivo** [comment]
I have the exactly same issue and watched the same pattern. I also had the same issue when I do the refresh sometimes.

Currently, I'm trying to remove some dependencies to check if there's some incompatibility. Once you have published your dependencies I'll try to remove the common dependencies between us.

**Comment by pedivo** [comment]
I've just found the problem. It's a conflict between OneSginal and IAP. They can't live together.
I haven't yet found a solution but I still working on that.

**Comment by pedivo** [comment]
If you downgrade your OneSignal plugin to 2.0.4, that will works.

I've also opened an issue on OneSignal repository, https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/140

**Comment by bboldi** [comment]
@pedivo Ok, I have tested it and can confirm that downgrading onesignal to 2.0.4 "fixes" the problem - Thank you! this solves it for now - hopefully this incompatibility will be fixed in future.

**Comment by bboldi** [comment]
This issue have been solved in release 2.0.11 of OneSignal https://github.com/OneSignal/OneSignal-Cordova-SDK/issues/140
